### PR TITLE
Launchpad: Show modal to return to Launchpad after the user saves a change in the site editor

### PR DESCRIPTION
**Pull request by sharpirate**
#### Summary
Currently, there's no easy way for the user to go back to Launchpad after they have made changes in the site editor. This PR creates a Modal that opens every time the user saves a change.

#### Testing instructions:
There are two possibilities for testing this - testing on the local jetpack docker setup or testing it on WPCOM by using `jetpack rsync`

##### Notes:
You should already have setup your local jetpack environment. I've [followed these instructions](https://github.com/Automattic/jetpack/blob/trunk/tools/docker/README.md#quick-install-wordpress) for the docker setup

##### Testing on WPCOM
1. Create a new Link in Bio site
2. Checkout this branch
3. Run `jetpack watch` and select `plugins -> jetpack`
4. Run `jetpack rsync` to sync your local changes to your sandbox ([Fieldguide page](PCYsg-eg0-p2))
5. Connect to your sandbox. You should see your changes applied.
6. Sandbox `public-api.wordpress.com` and your site URL
7. Go to `YOUR_SITE/wp-admin`
8. Go to the site editor (Appearance -> Editor)
9. Make a change in the site editor and click "Save"
10. You should see the Launchpad modal appear
11. Clicking "Back to Edit" or the close button should close the modal
12. Clicking "Launch Site" should take you to the Launchpad screen

**NOTE:** You might have to include `define( 'JETPACK_BETA_BLOCKS', true );` in your sandbox `wp-config.php`. This has been flaky for me, initially it didn't work without it but now it does so I'm not sure if this is required for an extension plugin...

https://user-images.githubusercontent.com/20927667/204922136-2b052594-66c8-40d4-919b-ee95d9edf106.mov

Fixes #69757

##### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

### CODE REVIEW

**Comment by Addison-Stavlo** [comment]
> Run jetpack watch and select plugins -> jetpack
Run jetpack rsync to sync

Also, `bin/jetpack-downloader test jetpack add/launchpad_on_save_modal` from this comment works as well for a little bit of an easier testing flow. - https://github.com/Automattic/jetpack/pull/27538#issuecomment-1322716144 

> NOTE: You might have to include define( 'JETPACK_BETA_BLOCKS', true ); in your sandbox wp-config.php. T

I was having issues at first so I did add this, but then realized my issue was a misspelled url in my hosts file. I think it would make sense if it worked without it on wpcom though.

**Review by Addison-Stavlo — COMMENTED** [review]
Very cool to see this starting to work! I am able to get this to show up when I save in the editor. 😁 

This seems to work with all the changes i make in the site editor so far. Note that this also gets triggered in the post editor too when editing a separate entity (such as a site title block, reusable block, or other similar item).

Screen Shot 2022-12-01 at 2 04 58 PM

Theres probably a decent way to determine if we are actually in the site editor, im happy to help brainstorm!

Other than that, it looks like theres no checking of site option / flow? I think you commented in slack that you might be looking into that currently?

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js
Comment (Addison-Stavlo) [inline projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js]: neat!

I imagine we will also need to check the specific flow, since the current copy is very specific to LIB (and i imagine we could follow up to add other variations for other flows). I think we were using a `site_intent` option or something like that?

Comment (jeyip) [reply projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js]: Second all of Addison's thoughts. I also think the CTA "Launch Site" is a little misleading, since all it does is redirect to Launchpad.

Maybe we could come up with a more context appropriate name like "Next Steps"

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php
Comment (jeyip) [inline projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php]: Can we use `add_inline_script` instead of `localize_script`?

Common consensus in the community is that we originally implemented `localize_script` because it _used_ to be the only way to hydrate client data from PHP. Now, `add_inline_script` exists, and `localize_script` should primarily be reserved for, well, localization.
https://developer.wordpress.org/reference/functions/wp_add_inline_script/

I realize that other plugins don't do this, but since this is a brand new file, I think it's pretty easy to avoid introducing new tech debt.

**Comment by jeyip** [comment]
### Requirements
- [x] When saving changes in the site editor, a launchpad modal appears
- [x] Clicking the close button closes the modal
- [x] Clicking the back to the editor button closes the modal
- [x] Clicking the launch site button redirects to Launchpad

### Browsers
- [x] Edge
- [x] Firefox
- [x] Chrome

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js
Comment (agrullon95) [reply projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js]: > Second all of Addison's thoughts. I also think the CTA "Launch Site" is a little misleading, since all it does is redirect to Launchpad.

+1 on this.

I got confused for a moment when I was initially testing because I thought it was going to immediately launch my site based on the text instead of sending me back to the LaunchPad screen.

> Maybe we could come up with a more context appropriate name like "Next Steps"

I think "Next Steps" should be sufficient here since its consistent with the calypso/wp-admin sidebar launchpad notice we just added and is not as confusing.

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php
Comment (sharpirate) [reply projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php]: Totally, it was supposed to be `add_inline_script`, I guess my changes got mixed up. I'll update that

Comment (mattwiebe) [reply projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php]: `wp_add_inline_script` is a slight improvement over `wp_localize_script` but you have to be sharp with escaping to ensure no CSRF attacks can happen. See an [example of it in action in the Site Editor](https://github.com/WordPress/WordPress/blob/8ea348f6dbfb3907a41a7bf9c92810966437c3eb/wp-admin/site-editor.php#L120-L128).

**Comment by mattwiebe** [comment]
This looks good and tests well.

One thing I'm wondering is if we want to tweak the behavior on multiple saves? I can't think of a great reason to not show it after every save, but I'm also thinking about a theoretical user who's going through and making iterative changes over their entire site, saving compulsively as they go so as to not lose their progress. This popup would be a big bummer for that user, with no way to make it go away after seeing it for the first time.

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php
Comment (mattwiebe) [inline projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/launchpad-save-modal.php]: If we stick with `wp_localize_script`, I would want to change the `currentSite` string here to something else. This is creating `window.currentSite` in the global JS namespace and could easily collide with another generically named global. So we usually pick something much less generic, prefix it with an underscore, or both.

So here I'd choose `LaunchPadModalOptions` or `_launchpadModalOptions` or something along those lines.

**Comment by Addison-Stavlo** [comment]
> One thing I'm wondering is if we want to tweak the behavior on multiple saves? I can't think of a great reason to not show it after every save, but I'm also thinking about ...
 
Its a good point to think about @mattwiebe !  I don't think its a blocker for getting an initial release out, but still important to consider (and something we could iterate on in the future). For now at least the modal is only temporary (until they launch), easy to dismiss (clicking outsdie, X button, keep editing button), etc. - it seems there is potential to be mildly annoying, but also hopefully supplying an outweighing level of usefulness.

Id be curious about ideas how to make it less annoying. Only showing it once would make it easy to get lost and not provide its value to a potentially large subset of the users, while setting it up to be minimizable or something else might take a bit more effort - but maybe its worth to do something like this in a next iteration? cc @nuriapenya as well for thoughts.

**Comment by sharpirate** [comment]
I've updated the modal to include a "Don't show again" checkbox as proposed by Nuria. 

image

I wanted to start a discussion about the best way to store the user selection of the checkbox such that it is persisted when the user refreshes the page or even after exiting the session. I was thinking we could use `localStorage` or we can save the selection as a setting on the BE if that's possible. I am open to suggestions...

**Comment by Addison-Stavlo** [comment]
> I wanted to start a discussion about the best way to store the user selection of the checkbox such that it is persisted when the user refreshes the page or even after exiting the session. I was thinking we could use localStorage or we can save the selection as a setting on the BE if that's possible. I am open to suggestions...

Lets chat about it in the team meeting today! There are pros/cons to the various approaches on both the user and dev side. Im also happy to have this updated in a follow up.

**Comment by Addison-Stavlo** [comment]
> Lets chat about it in the team meeting today! 

We did chat about this some in the meeting today, and are fine with it going the way it is currently (some more thoughts about it on the team meeting p2 - paYE8P-2iv-p2). We can always update that later too if it becomes an issue.

This is looking pretty good to me, but I'm going to add the jetpack-crew review label so we can get that part of the process and valuable input going 😁

**Comment by edanzer** [comment]
Just tested this. Modal shows and looks great when saving the post editor! Both buttons/links work. 

The extra checkbox to disable the modal does not seem to be working. I know @Sharpirate raised a question about how to store this, so perhaps this expected. In our meeting today, I think some were thinking we were storing and using this value in local component state already, but seems like we may not be. 

To follow up on @Addison-Stavlo's note above, I think we did decide it would be enough to store this in component state for now. That would mean that during any given editing session, you'd only see the modal once if you check that box. But if you reload the page, or navigate away and come back, you'd see it again until you dismiss again.

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js
Comment (Addison-Stavlo) [inline projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js]:
```suggestion
		const showModal =
			launchpadModalOptions.launchpadScreenOption === 'full' &&
			launchpadModalOptions.siteIntentOption === 'link-in-bio' &&
			isInsideSiteEditor &&
			isModalOpen &&
			! dontShowAgain;
```
It looks like we have `useState` and checkbox toggle both set up for `dontShowAgain/setDontShowAgain`. But the modal does show again, I imagine something similar to this suggestion might be what we were missing?

**Comment by Addison-Stavlo** [comment]
Thanks for fixing that last item @Sharpirate !  From my end this is looking good, but Im interested to hear what @Automattic/jetpack-crew thoughts are here - does this make sense in /plugins/jetpack if this is only intended for wpcom simple/atomic? etc. other uncertainties just since this repo is a bit less familiar to us.

**Review by Addison-Stavlo — DISMISSED** [review]
marking approved, but pending jetpack core team review here.

**Comment by agrullon95** [comment]
I tested again with the latest changes, but I'm still seeing the modal popup if I toggle (checked) `Don't show this again.`, click `Back to Edit`, make changes in the editor and save again. If I toggle (checked) `Don't show this again.` and click the `X (close)` button, the modal does not pop up the next time I save another change on the site editor.

@Addison-Stavlo - did you notice this same behavior for both `Back to Edit` and `X` (close) button after toggling the `Don't show this again.` checkbox?

**Comment by Addison-Stavlo** [comment]
> @Addison-Stavlo - did you notice this same behavior for both Back to Edit and X (close) button after toggling the Don't show this again. checkbox?

Oh nice catch!  No, shame on me: I missed that with the other button!

**Comment by sharpirate** [comment]
That was a nice catch indeed! I had to explicitly save the state upon clicking the "Back to Edit" button. I've updated the PR

**Review by agrullon95 — DISMISSED** [review]
> That was a nice catch indeed! I had to explicitly save the state upon clicking the "Back to Edit" button. I've updated the PR

Tested the modal close behavior again.
The modal does not open again after I toggle `Don't show this again.`, either click the `x` (close) icon, click the `Back to Edit` button, or click outside of the modal and then I save another change in the editor.

**Review by coder-karen — COMMENTED** [review]
> but Im interested to hear what https://github.com/orgs/Automattic/teams/jetpack-crew thoughts are here - does this make sense in /plugins/jetpack if this is only intended for wpcom simple/atomic?

👋 I decided to test this out simply by checking out the PR on a Jurassic Ninja site using the Jetpack Beta tester, to see if it would play well with Jetpack sites. With error logging enabled there were several errors printed out, and some dashboard pages wouldn't work at all as well (for example Appearance - Editor).

Here are some of the error notices (plus some 'headers already sent'  warnings):

```
Notice: Function wp_enqueue_script was called incorrectly. Scripts and styles should not be registered or enqueued until the wp_enqueue_scripts, admin_enqueue_scripts, or login_enqueue_scripts hooks. This notice was triggered by the launchpad-save-modal handle. Please see Debugging in WordPress for more information. (This message was added in version 3.3.0.) in /srv/users/user14bb486e/apps/user14bb486e/public/wp-includes/functions.php on line 5835

Notice: Function wp_add_inline_script was called incorrectly. Scripts and styles should not be registered or enqueued until the wp_enqueue_scripts, admin_enqueue_scripts, or login_enqueue_scripts hooks. This notice was triggered by the launchpad-save-modal handle. Please see Debugging in WordPress for more information. (This message was added in version 3.3.0.) in /srv/users/user14bb486e/apps/user14bb486e/public/wp-includes/functions.php on line 5835
```



If the feature is only meant for wpcom / WoA then hiding the code behind a WoA / WPcom check would prevent a self-hosted Jetpack site from trying to run the code (for example with what was done here: https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php#L29).

In terms of where to put the code, there was an internal Slack conversation a month ago that might be helpful / relevant here too: p1667936779903679-slack-CDLH4C1UZ 

Hope this is useful - I'll keep this PR as 'needs review' too just in case more Crew eyes can take a look.

**Review by jeherve — CHANGES_REQUESTED** [review]
> Here are some of the error notices (plus some 'headers already sent' warnings):

This is definitely problematic, on self-hosted Jetpack but also on WordPress.com.

I see a few things we could change here:

1) We're hooking `wp_enqueue_script()` directly into `plugins_loaded`: https://github.com/Automattic/jetpack/blob/302157fff27330e7d848ae40529cdde5d8ab0020/projects/plugins/jetpack/class.jetpack.php#L659
Ideally, we'd only enqueue at the right moment, and only on the right page where we need the resource, maybe creating a new function that would contain `wp_enqueue_script ` and `wp_add_inline_script`, and hooking into `enqueue_block_assets`. In there we would bail if `wp_should_load_block_editor_scripts_and_styles()` is set to false.
2) We cannot enqueue that raw `index.js` file directly, we'd most likely need to serve a built version of it instead.

Given all that, and since the modal lives and is displayed in the editor, would it be an option to just forgo the whole custom enqueuing? I assume you set it up that way so you could add the inline script with the options? In this case, we could add the inline script to the existing Jetpack editor bundle, `jetpack-blocks-editor`, that gets enqueued in the editor. Another, possibly even simpler option, would be to add those options to the initial state that gets loaded in the editor:
https://github.com/Automattic/jetpack/blob/bc051609cc7972dc676ebbd0340ff7424bf40390/projects/plugins/jetpack/class.jetpack-gutenberg.php#L721-L725

Let me know if this helps a bit!

**Comment by sharpirate** [comment]
@coder-karen I really appreciate the feedback! Are you seeing these error logs on your wpcom sandbox as well? I wasn't able to reproduce them on mine..

**Comment by coder-karen** [comment]
> Are you seeing these error logs on your wpcom sandbox as well? I wasn't able to reproduce them on mine..

No that was just testing using the Jetpack Beta plugin on a Jurassic Ninja site to replicate how a Jetpack self-hosted site would behave with this PR.

No errors on my sandbox though, works well there!

**Comment by sharpirate** [comment]
@coder-karen Thanks for confirming! I was able to see the errors on Jurassic Ninja as well. I've updated the way I enqueue the scripts and not I don't see them. I would appreciate a re-test

**Comment by coder-karen** [comment]
@Sharpirate I am still seeing warnings and notices but not related to scripts. Specifically the following:

```
Notice: Constant FEATURE_NAME already defined in /srv/users/user42d580ec/apps/user42d580ec/public/wp-content/plugins/jetpack-dev/extensions/plugins/post-publish-qr-post-panel/post-publish-qr-post-panel.php on line 9

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user42d580ec/apps/user42d580ec/public/wp-content/plugins/jetpack-dev/extensions/plugins/post-publish-qr-post-panel/post-publish-qr-post-panel.php:9) in /srv/users/user42d580ec/apps/user42d580ec/public/wp-admin/includes/misc.php on line 1416
```

Plus a few more headers already sent warnings, all suggesting headers are already sent in `qpost-publish-qr-post-panel.php`.

They all appear to be related to the `FEATURE_NAME` const defined on line 9 of `launchpad-save-modal.php`, as the same const name is used in `post-publish-qr-post-panel.php` (it seems also to be used in the payments plugin `payments.php` but that isn't causing any conflicts for some reason). If I change the name of that const in the launchpad file, the warnings and notice disappear for me.

**Comment by sharpirate** [comment]
@coder-karen I wasn't sure if those were related to the PR but that makes sense. I updated the feature name and the warnings disappeared. Thanks for pointing that out!

**Comment by edanzer** [comment]
@Sharpirate - This is testing great for me now!

- [x] If I create a new site via a flow, going to the site editor, make changes, and save, this modal shows. 
- [x] As expected, it will keep showing if I keep saving without ticking the 'do not show again' box.
- [x] If I tick the 'do not show again box', then the modal will stop showing as long as I stay on the editor page. As expected, if I refresh the page or navigate away and back, the modal will start showing again. 
- [x] Both modal links work as expected. Back to Edit just keeps us on the current page. Launch Site takes us back to Launchpad. 
- [x] Once I do launch my site, then this modal stops showing entirely. 

I'm approving for my part - as a test of functionality. But a few notes: 
   - I'm not a seasoned Jetpack dev. I assume the JP crew will catch issues of concern around code for Jetpack specifically. And obviously we'd need to resolve the current issues they've flagged before merging. 
   - I see the main component inside [launchpad-save-modal/index.js](https://github.com/Automattic/jetpack/pull/27538/files#diff-13206839dd3315d528c316c3405fbf89039de38e648bc7b16ac786db91fbb37d) is named `settings`. This is a matter of taste, and I don't feel strongly, but seems like we could name that component more descriptively, like LaunchpadSaveModal or something similar.
   - I added a comment elsewhere, but right now it looks like we're just enabling this for link-in-bio. Given free flow has a similar step to visit the editor, I wonder if it makes sense to add that. Could easily do so in a separate PR if we decide we want it there. 
   - In [launchpad-save-modal.php](https://github.com/Automattic/jetpack/pull/27538/files#diff-cd93d305018e20b74151c32b0b66243d940a984ab1b2c42c6cc48717d377d1a1), there's a php code sniffer warning about `You must use "/**" style comments for a function comment`. I assume we'd also want to resolve that prior to merge. 
   - I wouldn't hold up the PR over this, but because it's on my mind from working on translations lately, I notice that [launchpad-save-modal/index.js](https://github.com/Automattic/jetpack/pull/27538/files#diff-13206839dd3315d528c316c3405fbf89039de38e648bc7b16ac786db91fbb37d) is using __() from core's i18n. I think the ideal would be to use useTranslate() from Calypso. As it's being used within a component, it could be a straightforward change.


Finally just a note that I tested via this method: 
>  On your sandbox, run bin/jetpack-downloader test jetpack add/launchpad_on_save_modal to get started.

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js
Comment (edanzer) [inline projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js:35]: Would it make sense to enable this modal for the Free Flow now as well?

**Review by edanzer — DISMISSED** [review]

File: projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js
Comment (sharpirate) [reply projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js:35]: Yes! But I think it would be easier to address this in a follow up PR since we'll need different content inside the modal

**Comment by sharpirate** [comment]
@edanzer Thanks for the comments! I addressed some of them:

>In [launchpad-save-modal.php](https://github.com/Automattic/jetpack/pull/27538/files#diff-cd93d305018e20b74151c32b0b66243d940a984ab1b2c42c6cc48717d377d1a1), there's a php code sniffer warning about You must use "/**" style comments for a function comment

I updated the PR to use /** style comment

> I wouldn't hold up the PR over this, but because it's on my mind from working on translations lately, I notice that [launchpad-save-modal/index.js](https://github.com/Automattic/jetpack/pull/27538/files#diff-13206839dd3315d528c316c3405fbf89039de38e648bc7b16ac786db91fbb37d) is using __() from core's i18n.

I updated the PR to use `import { useTranslate } from 'i18n-calypso';` However I must say I am a bit worried since looks like this is the first time ever we are using `useTranslate` from calypso in jetpack (couldn't find other places in the repo) so I'm not sure if there's a reason for this..

> I see the main component inside [launchpad-save-modal/index.js](https://github.com/Automattic/jetpack/pull/27538/files#diff-13206839dd3315d528c316c3405fbf89039de38e648bc7b16ac786db91fbb37d) is named settings

This is a bit confusing but it's because the [registerJetpackPlugin](https://github.com/Automattic/jetpack/blob/add/launchpad_on_save_modal/projects/js-packages/shared-extension-utils/src/register-jetpack-plugin.js#L11) function is expecting a settings object with a render method. And on my settings object I have the method name as [LaunchpadSaveModal](https://github.com/Automattic/jetpack/blob/add/launchpad_on_save_modal/projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js#L14). So I think [this is a convention](https://github.com/Automattic/jetpack/blob/add/launchpad_on_save_modal/projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/editor.js#L4) we have to use

**Comment by jeherve** [comment]
> I updated the PR to use `import { useTranslate } from 'i18n-calypso';` However I must say I am a bit worried since looks like this is the first time ever we are using `useTranslate` from calypso in jetpack (couldn't find other places in the repo) so I'm not sure if there's a reason for this..

We cannot do that I'm afraid. This would pull in additional dependencies and would not be compatible with the way the WordPress.org infrastructure and this plugin handle internationalization.

>  I think [this is a convention](https://github.com/Automattic/jetpack/blob/add/launchpad_on_save_modal/projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/editor.js?rgh-link-date=2022-12-16T07%3A30%3A24Z#L4) we have to use

That's correct.

Would it be okay if I created a new PR against yours to address the things I had mentioned [in my review last week](https://github.com/Automattic/jetpack/pull/27538#pullrequestreview-1213471926)? It may be easier to show you what I meant that way, and should address some of the problems we're seeing with the build right now.

**Comment by jeherve** [comment]
> Would it be okay if I created a new PR against yours to address the things I had mentioned [in my review last week](https://github.com/Automattic/jetpack/pull/27538#pullrequestreview-1213471926)? It may be easier to show you what I meant that way, and should address some of the problems we're seeing with the build right now.

I opened a new PR against `trunk` here: #27976

Let me know what you think about it!

**Comment by sharpirate** [comment]
FYI: I'll be closing this PR in favor of https://github.com/Automattic/jetpack/pull/27976. Let's move the discussion there

**Pull request closed without merging**
